drivers: CAN: Add RX timestamp for CAN frames

alexanderwachter opened the pull request:
This PR adds an optional receive timestamp. The timestamp is a hardware feature of most CAN controllers. The counter value increments for every bit time of the CAN-bus. For example, a bus configured for 1Mboud increases the value every 1 microsecond. The counter is 16 bit wide.

Commit:
can: Add RX timestamp to zcan_frame

This commit introduces a timestamp for received CAN frames.
The timestamp is optional and can be activated via Kconfig.

Signed-off-by: Alexander Wachter <alexander.wachter@student.tugraz.at>

Commit:
drivers: can: stm32: Implement timestamp for RX frames

This commit implements timestamps for receiving frames on
stm32 CAN driver.

Signed-off-by: Alexander Wachter <alexander.wachter@student.tugraz.at>

jukkar commented on drivers/can/can_stm32.c:
```diff
@@ -52,6 +52,9 @@ static void can_stm32_get_msg_fifo(CAN_FIFOMailBox_TypeDef *mbox,
 	msg->dlc = mbox->RDTR & (CAN_RDT0R_DLC >> CAN_RDT0R_DLC_Pos);
```
Typo in commit subject of "dricers: can: stm32: Implement timestamp for RX frames"
s/dricers/drivers/

jukkar commented on drivers/can/Kconfig:
```diff
@@ -36,6 +36,15 @@ config CAN_WORKQ_FRAMES_BUF_CNT
+	  This options enable a timestamp value of the CAN free running timer.
+	  The value is incremented every bit time and starts when the controller
+	  is initialized.
+	  The maximum value is hardware dependent (stm32: 16bit, mcux: 16bit).
```
Is this always 16 bit? If not, should we just use int in zcan_frame struct?

alexanderwachter replied:
The timestamp is usually there for TTCAN which always uses 16-bit. The CANcan controllers we have, only support 16-bit or no timestamp (MCP2515). If we introduce an int, we would need to provide the max-value. Providing the max-value would need a new API because external controllers could differ from internal.

alexanderwachter replied:
Updated the comment.

henrikbrixandersen committed:
drivers: can: flexcan: implement timestamp for RX frames

This commit implements timestamps for receiving frames on the
NXP MCUX FlexCAN CAN driver.

Signed-off-by: Henrik Brix Andersen <henrik@brixandersen.dk>

henrikbrixandersen commented:
I have added RX timestamp support for the NXP MCUX FlexCAN driver in https://github.com/henrikbrixandersen/zephyr/commit/bd331afa802ba8960db52f5d3af233f40cf92064

ioannisg commented on drivers/can/Kconfig:
```diff
@@ -36,6 +36,14 @@ config CAN_WORKQ_FRAMES_BUF_CNT
+	bool "Enable receiving timestamps"
+	depends on CAN_STM32
+	help
+	  This options enable a timestamp value of the CAN free running timer.
```
This option enables

ioannisg commented on include/drivers/can.h:
```diff
@@ -173,6 +173,13 @@ struct zcan_frame {
 	};
+#if defined(CONFIG_CAN_RX_TIMESTAMP)
+	/** Timer value of the CAN free running timer.
+	 * The timer is incremented every bit time and captuered at the start
```
captured

alexanderwachter replied:
Done. Thanks!

alexanderwachter commented:
@henrikbrixandersen your patch is applied.
Thanks.

erwango reviewed (approved):
+1 for the STM32 part

ioannisg commented:
Does this require back-porting @alexanderwachter ?

alexanderwachter commented:
@ioannisg no. It's a new feature.
